Add tcp-established to checkpoint/restore

adrianreber opened the pull request:
This adds the possibility to checkpoint and restore containers with established TCP connections. See third commit for details and examples. This PR also includes updates to the bash completion which was missing a few of the recently added checkpoint/restore features.

adrianreber commented:
See https://github.com/kubernetes-sigs/cri-o/pull/1926 for required conmon changes.

adrianreber commented:
FYI @rst0git

mheon commented:
Since this depends on conmon changes, I'll hold off on adding an approved tag until that merges

rhatdan commented on libpod/container_api.go:
```diff
@@ -833,8 +833,14 @@ func (c *Container) Refresh(ctx context.Context) error {
 	KeepRunning bool
+	// TCPEstablished tells the API to checkpoint a container
+	// even if it contains established TCP connections
+	TCPEstablished bool
```
Should this default to true?

adrianreber replied:
This is something I was also thinking about. The main reason to default to false is that if there are established TCP connections the user might want to explicitly tell to actually checkpoint and restore. Maybe the environment does not handle checkpointing and restoring established TCP connections well, because an external firewall rule is needed to forward the traffic to the container and thus the migration of TCP connections makes no sense.

So I am unsure, most of the time I think, yes it should default to true as it automatically takes care of everything, but maybe the user should explicitly ask for it to not end up in a situation where the established connection cannot be forwarded through the port forwarding correctly.

I was following CRIU's and runc's default but for convenience we could default it to true. If the established TCP connections can be migrated, it is a win for the user. If not, the migration should still work and the TCP stack needs to handle the reconnection.

TomSweeneyRedHat replied:
if CRIU and runc default to false, I'd vote to keep this one false too.

rst0git replied:
I agree that tcp-established should be set to false by default. A user can always add `tcp-established` to the CRIU's configuration file to change this behaviour.

rhatdan commented on docs/podman-container-checkpoint.1.md:
```diff
@@ -29,6 +29,11 @@ Instead of providing the container name or ID, checkpoint the last created conta
+**--tcp-established**
+
+Checkpoint a container with established TCP connections. If the checkpoint image actually contains
+established TCP connections, this options is also required during restore.
```
Default (False)

adrianreber commented:
Hmm, all checkpoint tests are failing, strange. It works locally. Need to figure what is going on.

adrianreber commented:
Ah, so I have an idea why the checkpoint tests are failing. Since today criu 3.11 is part of all Fedora branches and now a criu version is automatically installed which has all the features for Podman.

Up until now the container in container checkpoint tests were not run, because the CRIU version in the container was too old. Now it is new enough, but CRIU cannot handle to checkpoint a container running in another container. Need to try the test cases (Podman in Podman) locally.

adrianreber commented:
Pushed a test commit. See also https://github.com/containers/libpod/pull/1855

giuseppe commented:
> Pushed a test commit. See also #1855

the patch LGTM.  If it fixes the CI failures it is probably worth to move it to its own PR so to unblock other PRs as well

giuseppe commented:
the patch fixes the tests, could you please move it to a separate PR?

TomSweeneyRedHat commented on docs/podman-container-restore.1.md:
```diff
@@ -32,6 +32,13 @@ Restore all checkpointed containers.
+Restore a container with established TCP connections. If the checkpoint image actually contains
+established TCP connections, this options is required during restore. If the checkpoint image
+does not contain established TCP connections this options is ignored. Defaults to not restoring
+containers with estblished TCP connections.
```
"estblishes" to "established"

TomSweeneyRedHat commented on completions/bash/podman:
```diff
@@ -716,16 +716,22 @@ _podman_container_attach() {
+     --latest
+     -R
+     --leave-running
+     --tcp-established
```
This is such a long param.  Can we also use '-t' or will that get a little wonky as we use -t for other purposes in different commands?

adrianreber replied:
Yes, it is long, but I would prefer to not change it. My reasons for not changing is to keep the same parameter as CRIU and runc. Some options already differ between CRIU and runc which I think makes it more complicated than necessary. So my goal is to stay as close as possible to CRIU to avoid further confusion of the involved parameters. With TAB-completion it should be not too difficult.

There are also more TCP related parameters in CRIU (`--tcp-close` - just close all open TCP connections and `--skip-in-flight` to ignore TCP states other than established and open), so that there is at least another parameter which could be shortened to `-t`.

Having said all that, I can add `-t` if it would offer a better user interface.

rst0git commented:
Since this PR is adding support for `tcp-established` to enable checkpoint of established TCP connections, should also add support for `skip-in-flight` and `tcp-close` for completeness?

adrianreber commented:
> Since this PR is adding support for `tcp-established` to enable checkpoint of established TCP connections, should also add support for `skip-in-flight` and `tcp-close` for completeness?

Yes and no :grin: The reason to not add it is to have only 'one' change in this PR and the real reason is, that runc does not yet support those CRIU flags. So first the changes need to be added to runc and then we can look at adding those features to Podman. Which makes more sense as another PR.

adrianreber commented:
@TomSweeneyRedHat thanks a lot for your review. I addressed all your reviews and force pushed a version.

Without the changes which are now part of #1858

adrianreber commented:
conmon changes have been merged

adrianreber commented:
/restest

mheon commented:
/approve

mheon commented:
/ok-to-test

mheon commented:
Let's see if the tests pick up the new conmon version or if we hardcode a commit somewhere

adrianreber commented:
The CI failures seem to be related to problems with the container registries.

adrianreber commented:
bot, retest this please

mheon commented:
Cirrus timeout is suspicious - I re-triggered, but we'll see.

Really annoying it's not actually throwing whatever error caused it to hang, if it is one.

mheon commented:
@adrianreber Seems to be timing out trying to restore?

`Running: /usr/bin/podman --root /tmp/podman_test449995664/crio --runroot /tmp/podman_test449995664/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --storage-driver=vfs container restore -l`

adrianreber commented:
@mheon thanks for looking into it. It seems my test for testing checkpointing with tcp-established does not work correctly in the CI. I am starting a container with --network=host and I am trying to connect to it. It works locally, but probably not in the CI. I will remove the test case to see if that helps. Too many changes in one PR it seems.

adrianreber commented:
I also see errors like this in cirrus https://cirrus-ci.com/task/6388882125881344:
```
varlink.error.VarlinkError: {'parameters': {'reason': 'cgroups: cgroup mountpoint does not exist'}, 'error': 'io.podman.ErrorOccurred'}```

Is this a known problem? Or triggered by my changes?

mheon commented:
Known flake that seems to have gotten a lot more common now - previously was 1 in 10 or 15 runs, now seems like it's happening every time. I'll grab @jwhonce and see if I can get a fix in.

adrianreber commented:
Cool, everything green. This includes an updated conmon and new CRIU version. Instead of pointing to the CRIU development branch which will be rebased/rewritten occasionally this now points to the stable branch which has stable commits.

From my point of view this is no longer WIP as the necessary conmon changes have been merged and pulled in in this PR for testing.

rhatdan commented:
LGTM
@mheon @baude @umohnani8 @jwhonce @vrothberg PTAL

mheon commented:
Looks fine, merging
/lgtm
